oauth2l-go/src/oauth2client/oauth2client.go

```diff
@@ -0,0 +1,316 @@
+// signs.
+func base64Encode(b []byte) string {
+	return strings.TrimSuffix(base64.URLEncoding.EncodeToString(b), "==")
+}
+
+// Creates a signed JWT token. Used for two-legged oauth token request.
+func createJWT(secret map[string]interface{}, scopes string, signer signerInterface) (string, error) {
+	// A valid JWT has an "iat" timestamp and an "exp" timestamp. Get the current
+	// time to create these timestamps.
+	now := int(time.Now().Unix())
+
+	// Construct the JWT header, which contains the private key id in the service
+	// account secret.
+	header := map[string]string{
+		"typ": "JWT",
+		"alg": "RS256",
+		"kid": secret["private_key_id"].(string),
+	}
+
+	// Construct the JWT payload.
+	payload := map[string]string{
+		"aud":   secret["token_uri"].(string),
+		"scope": scopes,
+		"iat":   strconv.Itoa(now),
+		"exp":   strconv.Itoa(now + 3600),
+		"iss":   secret["client_email"].(string),
+	}
+
+	// Convert the header and payload map to json.
+	headerJson, err := json.Marshal(header)
+	if err != nil {
+		return "", fmt.Errorf("Failed to convert JWT header to json.")
+	}
+	payloadJson, err := json.Marshal(payload)
+	if err != nil {
+		return "", fmt.Errorf("Failed to convert JWT payload to json.")
+	}
+
+	// Convert the header and payload json to base64
+	headerB64 := base64Encode(headerJson)
+	payloadB64 := base64Encode(payloadJson)
+
+	// The first two segments of the JWT are signed. The signature is the third
+	// segment.
+	segments := headerB64 + "." + payloadB64
+
+	// sign the hash, instead of the actual segments.
+	hashed := sha256.Sum256([]byte(segments))
+	signedBytes, err := signer.sign(rand.Reader, hashed[:], crypto.SignerOpts(sha256Opts{}))
+	if err != nil {
+		return "", fmt.Errorf("Failed to sign JWT: %s", err)
+	}
+
+	// Generate the final JWT as
+	// base64(header) + "." + base64(payload) + "." + base64(signature)
+	return segments + "." + base64Encode(signedBytes), nil
+}
+
+// Struct for OAuth2 Client
+type Client struct {
+	// The OAuth flow to run based on secret type.
+	flow func(scopes string) (*Token, error)
+
+	// The function to handle two-legged oauth authorization flow. This function
+	// takes in a authorization URL. After the user authorizes on the given url,
+	// returns the verification code.
+	authorizeHandler func(string) string
+}
+
+// Create a new OAuth2 Client
+// secretBytes: JSON text that represents either an OAuth client ID or a
+// service account.
+func NewClient(secretBytes []byte) (*Client, error) {
+	c := new(Client)
+	var secret map[string]interface{}
+	if err := json.Unmarshal(secretBytes, &secret); err != nil {
+		return nil, fmt.Errorf("Failed to parse secret json: %s", err)
+	}
+
+	if installed, found := secret["installed"]; found {
+		// When the secret contains "installed" field, it is a client secret. We
+		// will run a three-legged flow
+		installedMap := installed.(map[string]interface{})
+		c.flow = func(scopes string) (*Token, error) {
```

Comment:
While this logic works, It seems too confusing to read the code. I suggest we define an interface like this one: 

type TokenSource interface {
    // Token returns a token or an error.
    // Token must be safe for concurrent use by multiple goroutines.
    // The returned Token must not be modified.
    Token() (*Token, error)
}

Then have separate implementation for 2LO and 3LO. It will be easier for people to migrate from oauth2 to our oauth2client.